### ISSUE TITLE
ONNX import: strict symbolic-shape unification rejects common LLM decode exports

### ISSUE THREAD

**Issue by czoli1976** [sonos/tract#2343]
Surfaced while running a real int4 LLM through tract (validating #2340). Filing as a
scope/direction question with a self-contained repro — happy to take it on if there's appetite. @kali

## What happens
A standard optimum / transformers.js decode export declares `attention_mask` as
`[batch, past_sequence_length + 1]` (which pins the new sequence length to 1) but leaves
`input_ids` as `[batch, sequence_length]` — a free symbol. The KV `Concat` then produces
`past + sequence_length`, while the mask path implies `past + 1`, and tract can't unify
`sequence_length` with `1`:

```
Impossible to unify Add([Sym(past_sequence_length), Val(1)])
               with Broadcast([Add([Sym(sequence_length), Sym(past_sequence_length)]),
                                Add([Sym(past_sequence_length), Val(1)])]).
```

ONNXRuntime doesn't unify symbolically, so it just accepts it. tract is arguably *correct* to
reject — but it's a **very common export shape**, so many real int4/fp16 LLM ONNX models won't
import as-is.

## Minimal repro (no weights, no download)

<details><summary><code>make_repro.py</code> — a self-attention fragment with the offending mask shape</summary>

```python
import onnx, numpy as np
from onnx import TensorProto as T, helper as h
B, S, P = "batch_size", "sequence_length", "past_sequence_length"
vi = lambda n, dims: h.make_tensor_value_info(n, T.FLOAT, dims)
nodes = [
    h.make_node("Concat", ["past_k", "cur_k"], ["k"], axis=2),     # [B,2,P+S,4]
    h.make_node("Transpose", ["k"], ["kt"], perm=[0, 1, 3, 2]),    # [B,2,4,P+S]
    h.make_node("MatMul", ["q", "kt"], ["scores"]),                # [B,2,S,P+S]
    h.make_node("Unsqueeze", ["attention_mask", "ax"], ["m1"]),
    h.make_node("Unsqueeze", ["m1", "ax"], ["m2"]),                # [B,1,1,P+1]
    h.make_node("Add", ["scores", "m2"], ["out"]),                 # forces P+S == P+1
]
g = h.make_graph(nodes, "repro",
    [vi("q", [B,2,S,4]), vi("past_k", [B,2,P,4]), vi("cur_k", [B,2,S,4]),
     vi("attention_mask", [B, "past_sequence_length + 1"])],          # <-- the culprit
    [vi("out", [B,2,S,"past_sequence_length + 1"])],
    [onnx.numpy_helper.from_array(np.array([1], dtype="int64"), name="ax")])
onnx.save(h.make_model(g, opset_imports=[h.make_opsetid("",17)], ir_version=9), "repro.onnx")
```
</details>

```console
$ python make_repro.py && tract repro.onnx -f onnx dump
... Failed analyse for node "out" Add
... Impossible to unify Add([Sym(past_sequence_length), Val(1)]) with Broadcast(...)
```

## Real-model confirmation
`onnx-community/Qwen2.5-0.5B-Instruct` → `onnx/model_q4.onnx` (24-layer int4 decode model, 168
`MatMulNBits`) fails identically on import. After the workaround below it runs end-to-end and the
decode logits are **bit-identical to ONNXRuntime** (and all 168 int4 layers route through #2340's
Q4_0 path) — so the only blocker is the shape analysis, not any op.

## Workaround (the manual surgery this issue is about)

<details><summary><code>concretize.py</code> — set symbols concrete + strip <code>value_info</code> so tract re-infers</summary>

```python
import onnx, sys
m = onnx.load(sys.argv[1]); out = sys.argv[2]
syms = {k:int(v) for k,v in (kv.split("=") for kv in sys.argv[3:])}
def ev(e):
    e = e.strip()
    if e.lstrip("-").isdigit(): return int(e)
    if "+" in e: a,b = e.split("+"); return ev(a)+ev(b)
    return syms[e]
for vi in m.graph.input:
    for d in vi.type.tensor_type.shape.dim:
        if d.dim_param: d.dim_value = ev(d.dim_param); d.ClearField("dim_param")
del m.graph.value_info[:]
for o in m.graph.output: o.type.tensor_type.shape.Clear()
onnx.save(m, out)
```
</details>

```console
$ python concretize.py repro.onnx repro_concrete.onnx batch_size=1 sequence_length=1 past_sequence_length=0
$ tract repro_concrete.onnx -f onnx --onnx-ignore-output-shapes dump   # imports fine
```

## Question
Is there appetite for importing these exports more gracefully? Options:
1. An input-shape binding / `--concretize` that propagates symbol values *before* analyse (today,
   `-i` on one input conflicts with the others that still carry the symbols, so it has to be all-or-nothing).
2. Recognize the `[past + 1]` mask idiom as a decode export (`sequence_length = 1`).
3. Just document a concretization recipe and treat the export as inconsistent (user fixes the export).

Or out of scope by design? Happy to send a PR for whichever direction you prefer.
